Initial version

pushred opened the pull request:
OK, finally got around to tests. Added coverage for context mapping, routing, defaults, and also the tags Handlebars template. Addressed a number of issues that I discovered in that process. With that all in place I also added a few features that have been on the list and requested:
- Array items in context data can be accessed with whatever syntax feels right
- Context data can be interpolated into a string to help make sense of some of that data and also supporting the use of multiple properties
- Static routes are prioritized first, then dynamic routes
- `*` fallback routes can be specified and used to set defaults, those with more path segments are prioritized first

With all this, and the fact that it's already running in production, I'm ready to release version 1.0.0!

But please have a look first @joanniclaborde and @josiahsprague

pushred committed:
Modularize metadata preprocessor code

pushred committed:
Handle missing titles

joanniclaborde committed:
Add Handlebars template

pushred committed:
Add tests

pushred committed:
Test tags

pushred committed:
Support URLs from context data

pushred committed:
Add fallback routing

pushred committed:
Simplify fallback route syntax

Makes syntax consistent with auth config and glob paths

pushred committed:
Default metadata object to false when no routes matched

pushred committed:
Test defaults

pushred committed:
Support Handlebars/JS-style array selection

pushred committed:
Add context data string interpolation

Closes #2

localjo commented:
Seems like syntax errors in metadata.yml blow up the preprocessors with the only visible error being the first thing that tries to use preprocessed data that doesn't exist. Example;

```
TypeError: Cannot read property 'pages' of undefined
    at Object.module.exports (/vagrant/thecatsfoundation/node_modules/jungle-solidus/widgets/photo/default_options.js:2:68)
```

Not sure if there's a better way to surface errors here. But I think site devs are going to have a hard time figuring out what's going on if they make a syntax error in metadata.yml.

I also had trouble using [optional segments](http://millermedeiros.github.io/crossroads.js/#crossroads-add_route). We can work around that, we'll just have to be a little more verbose in metadata.yml.

localjo commented:
I've got a page that has something like `description: resources.story.0.excerpt`, but `excerpt` contains html tags, so it would be best to convert it to plain text first. Is there a way to process this via metadata.yml, or do we need to create a preprocessor to do those kinds of operations?

localjo commented:
Other than the comments above this is working smoothly for me. 👍

pushred committed:
Strip markup from context data

joanniclaborde commented on index.js:
```diff
@@ -0,0 +1,126 @@
+var urlRegex = require('url-regex');
+
+function parse( content, context ){
+  if( !content ) return false;
```
https://github.com/SparkartGroupInc/broadway/issues/6

joanniclaborde commented on index.js:
```diff
@@ -0,0 +1,126 @@
+
+  // Sort routes to prioritize deeper routes
+  var routes = Object.keys(metadata).sort(function( a, b ){
+    return b.match(/\//g).length > a.match(/\//g).length
```
Here you can just do `return b.match(/\//g).length - a.match(/\//g).length;`

joanniclaborde replied:
You could also match static routes first, else they get overridden by dynamic routes. Or we let the user deal with routing priority completely?

joanniclaborde replied:
Oh wait it's done below, never mind...

pushred replied:
Hmm compares might be [faster in V8](http://jsperf.com/comparison-vs-subtraction-sort), is a subtract sort better generally?

joanniclaborde replied:
Not really, it's just faster to read, and fits on one line. Pick your favorite!

joanniclaborde commented on index.js:
```diff
@@ -0,0 +1,126 @@
+
+  // Files
+  if( content.split('.').pop().match(/gif|jpg|jpeg|png/) ){
+    return content
```
How can `content` be empty?

pushred replied:
Good question! I'm thinking this was a lame attempt at error handling in the first pass. I've removed it, tests pass.

joanniclaborde commented on index.js:
```diff
@@ -0,0 +1,126 @@
+  }
+
+  // URLs
+  if( content && content.slice(0,4) === 'http' ){
```
You can remove the first condition here too, because of the first line of the function.

pushred committed:
Add error handling

pushred commented:
@josiahsprague Thanks for bringing up error handling, it was really entirely absent. Inexcusable! However syntax errors in your configuration isn’t something this module can check for. You either pass an object or not for the `metadata` argument, any issues parsing will be throwing exceptions from your code. I’ve added checks for this and a valid Solidus context. I also added some checks for every route pattern, which also addresses your optional segments issue.

I totally sympathize on the parsing exceptions, so I dug into that a bit. If you use a JSON file you at least get an error:

```
[2015-02-13T02:49:32.738Z] 7 Preprocessor Error:
SyntaxError: /example.com/preprocessors/metadata.json: Unexpected token :
    at Object.parse (native)
    at Object.Module._extensions..json (module.js:482:27)
    at Module.load (module.js:356:32)
    at Function.Module._load (module.js:312:12)
    at Module.require (module.js:364:17)
    at require (module.js:380:17)
    at Object.<anonymous> (/example.com/preprocessors/index.js:7:16)
    at Module._compile (module.js:456:26)
    at Object.Module._extensions..js (module.js:474:10)
    at Module.load (module.js:356:32)
```

But no line number! Turns out this is [an old V8 issue](https://code.google.com/p/v8/issues/detail?id=1480) that has been addressed at least as of 4.2.0. Even io.js isn’t running that version yet however. So it’s going to be awhile before handling is improved.

js-yaml is slightly more helpful:

```
YAMLException: JS-YAML: can not read a block mapping entry; a multiline key may not be an implicit key at line 22, column 8:
      title: Example
```

Unfortunately these exceptions are not being passed by the preprocessor workers, so instead we get this:

```
[2015-02-13T02:50:15.838Z] 9 Preprocessor Error:
undefined
```

I’ve filed https://github.com/solidusjs/solidus/issues/127 to investigate this further.

Thanks also for pointing out markup stripping, I actually had that in an earlier version and inadvertently removed it! It’s been restored.
